### ISSUE TITLE
Investigate initialization of NullLB WRT thread safety

### ISSUE THREAD

**Issue by PhilMiller** [UIUC-PPL/charm#887]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/887*

---
~~https://charm.cs.illinois.edu/gerrit/429~~ https://github.com/UIUC-PPL/charm/commit/be9a0cef3e724e97a3abbc5bf19cd8ca53670cb5
In that change, I had lingering concerns as follows:
> The NullLB fix isn't quite right, since other threads are still going to read from _theNullLB, and they'll just see a zero group ID, which causes a crash. I guess that assignment should be locked rather than gated on rank == 0.

From #538 comments.

**Comment by kavithachandrasekar** [comment]
*Original date: 2017-04-13 22:25:30*

---
I am trying to replicate the error. For some reason the init method is always called on other PEs after the NullLB::init method finishes on PE0, so the race never happens.

**Comment by PhilMiller** [comment]
*Original date: 2017-05-04 15:53:49*

---
I think the conclusion in the Core meeting was that this could be deferred due to limited impact and low likelihood of occurrence. It's not in the minutes, though, so I can't be sure. If that matches someone else's recollection, please change the Target Version to 6.8.1 or 6.9.0 as appropriate.

**Comment by kavithachandrasekar** [comment]
*Original date: 2017-05-04 16:32:22*

---
Yes, we did talk about deferring the bug. I will take a look at it today and if I cannot simulate the race, I will update the target version to 6.8.1.

**Comment by kavithachandrasekar** [comment]
*Original date: 2017-05-05 18:43:53*

---
When Group constructors are invoked on PE 0, broadcast messages are sent to other PEs to create the Group's chares on those PEs. However, the _initHandler on other PEs does not process these messages and buffers them till it has received all the initial messages related to groups, nodegroups, readonly data from PE 0. As a result, PE0 proceeds forward with more Group constructors before other PEs can call group chare constructors. Hence, NullLB constructor is always called on PE0 before other PEs.
However, as suggested before for correctness, I can change if(CmiMyRank() == 0) to a lock around _theNullLB = thisgroup. Please let me know.

**Comment by PhilMiller** [comment]
*Original date: 2017-05-05 18:52:31*

---
Knowing there's synchronization protecting this fully seems to be sufficient. If others agree, this can be closed.

Thank you for the investigation.
